src/utils/typedData.ts

```diff
@@ -107,130 +152,183 @@ function getMerkleTreeType(types: TypedData['types'], ctx: Context) {
+      });
+      return `${esc(dependency)}(${dependencyElements})`;
     })
     .join('');
-};
+}
 
 /**
  * Get a type string as hash.
  */
-export const getTypeHash = (types: TypedData['types'], type: string): string => {
-  return getSelectorFromName(encodeType(types, type));
-};
+export function getTypeHash(
+  types: TypedData['types'],
+  type: string,
+  revision: Revision = Revision.Legacy
+): string {
+  return getSelectorFromName(encodeType(types, type, revision));
+}
 
 /**
  * Encodes a single value to an ABI serialisable string, number or Buffer. Returns the data as tuple, which consists of
  * an array of ABI compatible types, and an array of corresponding values.
  */
-export const encodeValue = (
+export function encodeValue(
   types: TypedData['types'],
   type: string,
   data: unknown,
-  ctx: Context = {}
-): [string, string] => {
+  ctx: Context = {},
+  revision: Revision = Revision.Legacy
+): [string, string] {
   if (types[type]) {
-    // eslint-disable-next-line @typescript-eslint/no-use-before-define
-    return [type, getStructHash(types, type, data as Record<string, unknown>)];
+    return [type, getStructHash(types, type, data as TypedData['message'], revision)];
   }
 
-  if (
-    Object.keys(types)
-      .map((x) => `${x}*`)
-      .includes(type)
-  ) {
-    const structHashes: string[] = (data as unknown[]).map((struct) => {
-      // eslint-disable-next-line @typescript-eslint/no-use-before-define
-      return getStructHash(types, type.slice(0, -1), struct as Record<string, unknown>);
-    });
-    return [type, computeHashOnElements(structHashes)];
+  if (presetTypes[type]) {
+    return [type, getStructHash(presetTypes, type, data as TypedData['message'], revision)];
   }
 
-  if (type === 'merkletree') {
-    const merkleTreeType = getMerkleTreeType(types, ctx);
-    const structHashes: string[] = (data as unknown[]).map((struct) => {
-      // eslint-disable-next-line @typescript-eslint/no-use-before-define
-      return encodeValue(types, merkleTreeType, struct as Record<string, unknown>)[1];
-    });
-    const { root } = new MerkleTree(structHashes as string[]);
-    return ['felt', root];
-  }
+  if (type.endsWith('*')) {
+    const isStructArray = Object.keys(types)
+      .map((x) => `${x}*`)
+      .includes(type);
 
-  if (type === 'felt*') {
-    return ['felt*', computeHashOnElements(data as string[])];
-  }
+    const mappingMethod: (entry: Record<string, unknown>) => string = isStructArray
+      ? (entry) => getStructHash(types, type.slice(0, -1), entry, revision)
+      : (entry) => encodeValue(types, type.slice(0, 1), entry, undefined, revision)[1];
 
-  if (type === 'selector') {
-    return ['felt', prepareSelector(data as string)];
+    const hashes: string[] = (data as Array<TypedData['message']>).map(mappingMethod);
+    return [type, revisionConfiguration[revision].hashMethod(hashes)];
   }
 
-  return [type, getHex(data as string)];
-};
+  switch (type) {
```

Comment:
basic types like felt should no be overridden by the provided types

Reply:
Plan is to have it checked when the validation is reworked.

Reply:
IMO its easy to accomplish now, we just reorder to:
- check if preset
- check if array
- check other basic types
- check user defined types

and we can later add the extra validation with nice error message, but at least the priorities will be correct from the beginning

Reply:
It would be a change for the existing behaviour, it could be split by revision but I want to avoid adding too many of those conditionals. Idea is to get the correct hash in this version and enforce rules in the following version.